### PR TITLE
chore: fix ESLint not finding the tsconfig.json files

### PR DESCRIPTION
**Pull request by bapmrl**
## Context

Linear issue: [SM-962 [Tech debt] Spike 1 day - Properly integrate ESLint into VSCode](https://linear.app/prismic/issue/SM-962/[tech-debt]-spike-1-day-properly-integrate-eslint-into-vscode).

## The Solution

1. Use `.eslintrc.js` files instead of `.eslintrc.json` files so that we can pass `__dirname` to `parserOptions.tsconfigRootDir`.
2. List all `tsconfig.json` files in the `.eslintrc.js` files so that ESLint always has type information.
3. Add missing `tsconfig-node.json` file so that ESLint doesn't fail when trying to lint the `.eslintrc.js` files.
4. Add missing `"next/babel"` to `e2e-projects/next/.eslintrc.js`.
5. Add missing `tsconfig.json` file in `packages/client/__tests__`.
6. Set `compilerOptions.rootDir` to `".."` in `tsconfig.json` files located in `__tests__` folders.
7. Remove unrecommended `parserOptions.createDefaultProgram` in all `.eslintrc.js` files.
8. Remove useless `references` in all `tsconfig.json` files (as we're not using the `composite` option).

## Checklist before requesting a review
- [x] I hereby declare my code ready for review.
- [ ] If it is a critical feature, I have added tests.
- [x] The CI is successful.
- [ ] If there could backward compatibility issues, it has been discussed and planned.

## Preview

Screenshot 2023-01-18 at 12 20 27

### CODE REVIEW

File: packages/client/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  extends: [
+    "plugin:@typescript-eslint/recommended",
+    "plugin:@typescript-eslint/recommended-requiring-type-checking",
+    "plugin:jest/recommended",
+    "eslint-config-prettier",
+  ],
+  ignorePatterns: ["build"],
+  rules: {
+    "jest/no-mocks-import": "off",
+    "jest/no-conditional-expect": "off",
```
Comment (MarcMcIntosh) [inline packages/client/.eslintrc.js:26]: why `no-conditional-expect`?

File: packages/slice-machine/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  plugins: ["@typescript-eslint", "import", "jest"],
+  extends: [
+    "plugin:react-hooks/recommended",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:@typescript-eslint/recommended-requiring-type-checking",
+    "plugin:jest/recommended",
+    "eslint-config-prettier",
+  ],
+  ignorePatterns: ["build", "templates", "**/tests/**", "helpers/**"],
```
Comment (MarcMcIntosh) [inline packages/slice-machine/.eslintrc.js:25]: no linting on tests or helpers?

File: tsconfig-node.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "compilerOptions": { "noEmit": true },
+  "include": ["packages/*/.eslintrc.js"]
```
Comment (MarcMcIntosh) [inline tsconfig-node.json:3]: isn't this a circular reference?

File: packages/client/.eslintrc.js (hunk 1 above)
Comment (bapmrl) [reply packages/client/.eslintrc.js:26]: It was already specified in the initial `.eslintrc.json` file and I haven't changed it ^^

File: packages/slice-machine/.eslintrc.js (hunk 2 above)
Comment (bapmrl) [reply packages/slice-machine/.eslintrc.js:25]: Same thing, I haven't changed that and it's not the goal of the Linear issue I'm tackling.

File: tsconfig-node.json (hunk 3 above)
Comment (bapmrl) [reply tsconfig-node.json:3]: Why? `typescript-eslint` needs to know of the various `tsconfig.json` files but those files can reference/include any JavaScript/TypeScript source code.

Comment (MarcMcIntosh) [reply tsconfig-node.json:3]: like some of the "packages/*/.eslintrc.js" files also mention this one ::)

File: packages/client/.eslintrc.js (hunk 1 above)
Comment (MarcMcIntosh) [reply packages/client/.eslintrc.js:26]: 🤔

Comment (MarcMcIntosh) [reply packages/client/.eslintrc.js:26]: oh it's for io-ts Eithers :/

File: tsconfig-node.json (hunk 3 above)
Comment (bapmrl) [reply tsconfig-node.json:3]: Discussed on Slack 🙂

**Review by MarcMcIntosh — APPROVED** [review]

**Review by mathieud — APPROVED** [review]

**Pull request merged**